File: android/split/src/main/java/com/prince/split/SplitSheetsClient.java
package com.prince.split;

import androidx.annotation.Nullable;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Thin REST client for the bits of Sheets v4 + Drive v3 the SDK needs.
 *
 * <p>Auth is callers' problem — pass a fresh access token to every method. A 401 response
 * surfaces as {@link UnauthorizedException} so {@link SplitCloudSync} can re-auth and
 * retry without baking refresh logic into this layer.
 */
final class SplitSheetsClient {

    private static final String SHEETS_BASE = "https://sheets.googleapis.com/v4/spreadsheets";
    private static final int CONNECT_TIMEOUT_MS = 6000;
    private static final int READ_TIMEOUT_MS = 8000;

    /** Tab + columns the SDK uses; mirrors the legacy Apps Script schema. */
    static final String TAB = "Splits";
    private static final String[] HEADERS = {
            "timestampIso", "timestampMs", "deviceId", "amount", "people", "perPerson"
    };
    /** Columns A:F — full schema range for reads. */
    private static final String READ_RANGE = TAB + "!A2:F";

    static final class UnauthorizedException extends IOException {
        UnauthorizedException(String msg) { super(msg); }
    }

    static final class Row {
        final long timestampMs;
        final String deviceId;
        final double amount;
        final int people;
        Row(long timestampMs, String deviceId, double amount, int people) {
            this.timestampMs = timestampMs;
            this.deviceId = deviceId;
            this.amount = amount;
            this.people = people;
        }
    }

    private SplitSheetsClient() {}

    /**
     * Creates a new spreadsheet titled {@code "Turtle Splits"} with a {@code Splits} tab
     * and header row. Returns the new spreadsheet ID.
     */
    static String createSpreadsheet(String accessToken) throws IOException {
        JSONObject body = new JSONObject();
        try {
            JSONObject props = new JSONObject().put("title", "Turtle Splits");
            JSONObject sheet = new JSONObject().put("properties",
                    new JSONObject().put("title", TAB));
            body.put("properties", props)
                .put("sheets", new JSONArray().put(sheet));
        } catch (Exception e) {
            throw new IOException(e);
        }
        JSONObject resp = request("POST", SHEETS_BASE, accessToken, body.toString());
        String id = resp.optString("spreadsheetId", "");
        if (id.isEmpty()) throw new IOException("createSpreadsheet: empty spreadsheetId");
        // Stamp headers in row 1.
        appendRows(accessToken, id, Collections.singletonList(asStringRow(HEADERS)));
        return id;
    }

    /** Appends one or more value rows to the {@code Splits} tab. Each inner list is one row. */
    static void appendRows(String accessToken, String spreadsheetId, List<List<Object>> rows)
            throws IOException {
        appendRowsToTab(accessToken, spreadsheetId, TAB, rows);
    }

    private static void appendRowsToTab(String accessToken, String spreadsheetId,
                                        String tab, List<List<Object>> rows) throws IOException {
        if (rows.isEmpty()) return;
        String url = SHEETS_BASE + "/" + spreadsheetId + "/values/"
                + encode(tab + "!A1") + ":append?valueInputOption=RAW&insertDataOption=INSERT_ROWS";
        JSONArray values = new JSONArray();
        for (List<Object> row : rows) {
            JSONArray jsonRow = new JSONArray();
            for (Object cell : row) jsonRow.put(cell);
            values.put(jsonRow);
        }
        JSONObject body;
        try {
            body = new JSONObject().put("values", values);
        } catch (Exception e) {
            throw new IOException(e);
        }
        request("POST", url, accessToken, body.toString());
    }

    /** Owner-only: nukes every data row from the {@code Splits} tab while preserving the header. */
    static void deleteAllDataRows(String accessToken, String spreadsheetId) throws IOException {
        String url = SHEETS_BASE + "/" + spreadsheetId + "/values/"
                + encode(TAB + "!A2:F") + ":clear";
        request("POST", url, accessToken, "{}");
    }

    /** Reads every data row from the {@code Splits} tab into typed {@link Row}s. */
    static List<Row> listRows(String accessToken, String spreadsheetId) throws IOException {
        String url = SHEETS_BASE + "/" + spreadsheetId + "/values/" + encode(READ_RANGE);
        JSONObject resp = request("GET", url, accessToken, null);
        JSONArray values = resp.optJSONArray("values");
        if (values == null) return Collections.emptyList();
        List<Row> out = new ArrayList<>(values.length());
        for (int i = 0; i < values.length(); i++) {
            JSONArray r = values.optJSONArray(i);
            if (r == null || r.length() < 5) continue;
            try {
                long ts = parseLong(r.optString(1, "0"));
                String dev = r.optString(2, "");
                double amt = parseDouble(r.optString(3, "0"));
                int people = (int) parseDouble(r.optString(4, "0"));
                out.add(new Row(ts, dev, amt, people));
            } catch (NumberFormatException ignored) {}
        }
        return out;
    }

    /**
     * Deletes all data rows whose {@code deviceId} (column C) matches {@code deviceId}.
     * Iterates row indices from the bottom up so deletions don't shift indexes mid-loop.
     */
    static void deleteRowsForDevice(String accessToken, String spreadsheetId, String deviceId)
            throws IOException {
        // Need numeric sheetId for batchUpdate, not the title.
        int sheetId = resolveSheetId(accessToken, spreadsheetId, TAB);
        // Read all current values to find matching indexes.
        String url = SHEETS_BASE + "/" + spreadsheetId + "/values/" + encode(READ_RANGE);
        JSONObject resp = request("GET", url, accessToken, null);
        JSONArray values = resp.optJSONArray("values");
        if (values == null || values.length() == 0) return;

        // 0-based row indexes within the data range; absolute sheet rows are these + 1 (header).
        List<Integer> matches = new ArrayList<>();
        for (int i = 0; i < values.length(); i++) {
            JSONArray r = values.optJSONArray(i);
            if (r == null || r.length() < 3) continue;
            if (deviceId.equals(r.optString(2, ""))) matches.add(i + 1); // +1 = absolute row index (0-based with header at 0)
        }
        if (matches.isEmpty()) return;
        Collections.sort(matches, Collections.reverseOrder());

        JSONArray requests = new JSONArray();
        try {
            for (Integer idx : matches) {
                JSONObject delReq = new JSONObject()
                        .put("deleteDimension", new JSONObject()
                                .put("range", new JSONObject()
                                        .put("sheetId", sheetId)
                                        .put("dimension", "ROWS")
                                        .put("startIndex", idx)
                                        .put("endIndex", idx + 1)));
                requests.put(delReq);
            }
            JSONObject body = new JSONObject().put("requests", requests);
            request("POST", SHEETS_BASE + "/" + spreadsheetId + ":batchUpdate",
                    accessToken, body.toString());
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    private static int resolveSheetId(String accessToken, String spreadsheetId, String tabName)
            throws IOException {
        String url = SHEETS_BASE + "/" + spreadsheetId + "?fields=sheets.properties";
        JSONObject resp = request("GET", url, accessToken, null);
        JSONArray sheets = resp.optJSONArray("sheets");
        if (sheets == null) throw new IOException("no sheets in response");
        for (int i = 0; i < sheets.length(); i++) {
            JSONObject props = sheets.optJSONObject(i).optJSONObject("properties");
            if (props != null && tabName.equals(props.optString("title"))) {
                return props.optInt("sheetId", 0);
            }
        }
        throw new IOException("tab '" + tabName + "' not found");
    }

    // -- core HTTP --------------------------------------------------------------

    private static JSONObject request(String method, String url, String accessToken,
                                      @Nullable String jsonBody) throws IOException {
        HttpURLConnection conn = null;
        try {
            conn = (HttpURLConnection) new URL(url).openConnection();
            conn.setConnectTimeout(CONNECT_TIMEOUT_MS);
            conn.setReadTimeout(READ_TIMEOUT_MS);
            conn.setRequestMethod(method);
            conn.setRequestProperty("Authorization", "Bearer " + accessToken);
            conn.setRequestProperty("Accept", "application/json");
            if (jsonBody != null) {
                conn.setDoOutput(true);
                conn.setRequestProperty("Content-Type", "application/json; charset=utf-8");
                OutputStream os = conn.getOutputStream();
                os.write(jsonBody.getBytes("UTF-8"));
                os.flush();
                os.close();
            }
            int code = conn.getResponseCode();
            if (code == 401 || code == 403) {
                throw new UnauthorizedException("HTTP " + code + " from " + url);
            }
            if (code < 200 || code >= 300) {
                String err = readStream(conn.getErrorStream());
                throw new IOException("HTTP " + code + ": " + err);
            }
            String body = readStream(conn.getInputStream());
            return body.isEmpty() ? new JSONObject() : new JSONObject(body);
        } catch (IOException e) {
            throw e;
        } catch (Exception e) {
            throw new IOException(e);
        } finally {
            if (conn != null) conn.disconnect();
        }
    }

    private static String readStream(@Nullable java.io.InputStream is) throws IOException {
        if (is == null) return "";
        BufferedReader r = new BufferedReader(new InputStreamReader(is, "UTF-8"));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = r.readLine()) != null) sb.append(line);
        r.close();
        return sb.toString();
    }

    // -- helpers ----------------------------------------------------------------

    private static List<Object> asStringRow(String[] cols) {
        List<Object> row = new ArrayList<>(cols.length);
        Collections.addAll(row, cols);
        return row;
    }

    private static String encode(String s) {
        try {
            return URLEncoder.encode(s, "UTF-8");
        } catch (Exception e) {
            return s;
        }
    }

    private static long parseLong(String s) {
        try { return (long) Double.parseDouble(s); } catch (Exception e) { return 0L; }
    }

    private static double parseDouble(String s) {
        try { return Double.parseDouble(s); } catch (Exception e) { return 0.0; }
    }
}

File: ios/TurtleKeyboard/Cloud/SplitSheetsClient.swift
import Foundation

/// Thin REST client for the Sheets v4 surface the SDK needs. Mirrors
/// Android's `SplitSheetsClient` schema (tab name, header columns, row
/// shape) so a sheet created on one platform reads identically on the other.
///
/// Auth is the caller's problem — pass a fresh access token to every call.
/// 401/403 surfaces as `unauthorized` so `SplitCloudSync` can re-auth.
enum SplitSheetsClient {

    static let tab = "Splits"
    private static let headers = [
        "timestampIso", "timestampMs", "deviceId", "amount", "people", "perPerson",
    ]
    private static let readRange = "\(tab)!A2:F"
    private static let base = "https://sheets.googleapis.com/v4/spreadsheets"

    enum SheetsError: Error, LocalizedError {
        case unauthorized(String)
        case http(Int, String)
        case decode(String)

        var errorDescription: String? {
            switch self {
            case .unauthorized(let m): return "Auth: \(m)"
            case .http(let c, let m):  return "HTTP \(c): \(m)"
            case .decode(let m):       return "Decode: \(m)"
            }
        }
    }

    struct Row {
        let timestampMs: Int64
        let deviceId: String
        let amount: Double
        let people: Int
    }

    // MARK: - Public

    /// Creates a new spreadsheet titled "Turtle Splits" with a Splits tab
    /// and header row. Returns the new spreadsheet ID.
    static func createSpreadsheet(accessToken: String) async throws -> String {
        let body: [String: Any] = [
            "properties": ["title": "Turtle Splits"],
            "sheets": [["properties": ["title": tab]]],
        ]
        let resp = try await request("POST", url: base, accessToken: accessToken, body: body)
        guard let id = resp["spreadsheetId"] as? String, !id.isEmpty else {
            throw SheetsError.decode("no spreadsheetId")
        }
        // Stamp headers in row 1.
        try await appendRows(accessToken: accessToken,
                             spreadsheetId: id,
                             rows: [headers.map { $0 as Any }])
        return id
    }

    /// Appends one or more rows to the Splits tab.
    static func appendRows(
        accessToken: String,
        spreadsheetId: String,
        rows: [[Any]]
    ) async throws {
        guard !rows.isEmpty else { return }
        let url = "\(base)/\(spreadsheetId)/values/\(encode(tab + "!A1")):append"
            + "?valueInputOption=RAW&insertDataOption=INSERT_ROWS"
        let body: [String: Any] = ["values": rows]
        _ = try await request("POST", url: url, accessToken: accessToken, body: body)
    }

    /// Reads every data row from the Splits tab.
    static func listRows(accessToken: String, spreadsheetId: String) async throws -> [Row] {
        let url = "\(base)/\(spreadsheetId)/values/\(encode(readRange))"
        let resp = try await request("GET", url: url, accessToken: accessToken, body: nil)
        guard let values = resp["values"] as? [[Any]] else { return [] }
        var out: [Row] = []
        out.reserveCapacity(values.count)
        for r in values {
            guard r.count >= 5 else { continue }
            let ts: Int64 = parseLong(r[1])
            let dev = stringOf(r[2])
            let amt = parseDouble(r[3])
            let people = Int(parseDouble(r[4]))
            out.append(Row(timestampMs: ts, deviceId: dev, amount: amt, people: people))
        }
        return out
    }

    /// Owner-only: nukes every data row from the Splits tab while
    /// preserving the header.
    static func deleteAllDataRows(accessToken: String, spreadsheetId: String) async throws {
        let url = "\(base)/\(spreadsheetId)/values/\(encode(tab + "!A2:F")):clear"
        _ = try await request("POST", url: url, accessToken: accessToken, body: [:])
    }

    /// Deletes all data rows whose deviceId (column C) matches `deviceId`.
    /// Iterates row indices from the bottom up so deletions don't shift
    /// indexes mid-loop.
    static func deleteRowsForDevice(
        accessToken: String,
        spreadsheetId: String,
        deviceId: String
    ) async throws {
        let sheetId = try await resolveSheetId(accessToken: accessToken,
                                               spreadsheetId: spreadsheetId,
                                               tabName: tab)
        let listURL = "\(base)/\(spreadsheetId)/values/\(encode(readRange))"
        let resp = try await request("GET", url: listURL, accessToken: accessToken, body: nil)
        guard let values = resp["values"] as? [[Any]], !values.isEmpty else { return }

        var matches: [Int] = []
        for (i, r) in values.enumerated() {
            guard r.count >= 3 else { continue }
            if stringOf(r[2]) == deviceId {
                matches.append(i + 1) // +1 = absolute row index (header row is at 0)
            }
        }
        guard !matches.isEmpty else { return }
        matches.sort(by: >)

        let requests: [[String: Any]] = matches.map { idx in
            [
                "deleteDimension": [
                    "range": [
                        "sheetId": sheetId,
                        "dimension": "ROWS",
                        "startIndex": idx,
                        "endIndex": idx + 1,
                    ],
                ],
            ]
        }
        let body: [String: Any] = ["requests": requests]
        _ = try await request("POST",
                              url: "\(base)/\(spreadsheetId):batchUpdate",
                              accessToken: accessToken,
                              body: body)
    }

    // MARK: - Internals

    private static func resolveSheetId(
        accessToken: String,
        spreadsheetId: String,
        tabName: String
    ) async throws -> Int {
        let url = "\(base)/\(spreadsheetId)?fields=sheets.properties"
        let resp = try await request("GET", url: url, accessToken: accessToken, body: nil)
        guard let sheets = resp["sheets"] as? [[String: Any]] else {
            throw SheetsError.decode("no sheets in response")
        }
        for s in sheets {
            if let props = s["properties"] as? [String: Any],
               let title = props["title"] as? String, title == tabName,
               let id = props["sheetId"] as? Int {
                return id
            }
        }
        throw SheetsError.decode("tab '\(tabName)' not found")
    }

    private static func request(
        _ method: String,
        url: String,
        accessToken: String,
        body: Any?
    ) async throws -> [String: Any] {
        guard let u = URL(string: url) else {
            throw SheetsError.decode("bad URL")
        }
        var req = URLRequest(url: u)
        req.httpMethod = method
        req.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if let body = body {
            req.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else {
            throw SheetsError.http(0, "no response")
        }
        if http.statusCode == 401 || http.statusCode == 403 {
            let msg = String(data: data, encoding: .utf8) ?? ""
            throw SheetsError.unauthorized("HTTP \(http.statusCode): \(msg)")
        }
        guard (200..<300).contains(http.statusCode) else {
            let msg = String(data: data, encoding: .utf8) ?? ""
            throw SheetsError.http(http.statusCode, msg)
        }
        guard !data.isEmpty else { return [:] }
        let json = try JSONSerialization.jsonObject(with: data)
        return (json as? [String: Any]) ?? [:]
    }

    // MARK: - Cell helpers

    private static func encode(_ s: String) -> String {
        s.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? s
    }
    private static func stringOf(_ v: Any) -> String {
        if let s = v as? String { return s }
        return String(describing: v)
    }
    private static func parseLong(_ v: Any) -> Int64 {
        if let n = v as? NSNumber { return n.int64Value }
        if let s = v as? String, let n = Int64(s) { return n }
        if let s = v as? String, let d = Double(s) { return Int64(d) }
        return 0
    }
    private static func parseDouble(_ v: Any) -> Double {
        if let n = v as? NSNumber { return n.doubleValue }
        if let s = v as? String, let d = Double(s) { return d }
        return 0
    }
}
